File: Leetcode/rnnfeol.py
# leetcode no.19 Remove Nth Node From End of List

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        dummy, cur = ListNode(-1), head
        dummy.next = cur
        node_list = [dummy]
        while cur != None:
            node_list.append(cur)
            cur = cur.next

        prev, cur = node_list[-n-1], node_list[-n]
        prev.next = cur.next

        return dummy.next

File: Leetcode/rnnfeol.cpp
#include <iostream>
#include <vector>
using namespace std;
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* dummy = new ListNode();
		ListNode* slow = dummy, fast = dummy;
		dummy->next = head;
        for(int i = 0; i <= n; i++) fast = fast->next;
        
		while(fast != nullptr){
			fast = fast->next;
			slow = slow->next;
		}
		slow->next = slow->next->next;
		return dummy->next;
    }
};
